README
$Id: README 9522 2009-10-04 04:03:08Z FREQUENCY@cpan.org $

Alien::Libjio is a utility package to install and locate libjio

Copyright 2009 by Jonathan Yu <jawnsy@cpan.org>. All rights reserved.

INSTALLATION

  This module depends on Module::Build for its installation mechanism.

  To install this module type the following:
  $ perl Build.PL
  $ perl Build
  $ perl Build test
  $ perl Build install

  Module::Build will complain if any dependencies are missing, or if
  you can install some other modules to enhance its functionality.

USAGE AND DOCUMENTATION

  In order to use the module, you'll probably need the documentation. Once
  installed, you can read that using:
  $ perldoc Alien::Libjio

LICENSE

  In a perfect world, I could just say that this package and all of the
  code it contains is Public Domain. It's a bit more complicated than
  that; you'll have to read LICENSE to get the full details.

  Here are some nice things to live by:
  1. Give credit where it's due.
  2. Share your modifications, so everyone benefits from them.
  3. Do something nice for the authors.
  4. Help someone who needs it.
  5. Don't waste. Anything. Be especially careful not to waste energy that
     comes from non-renewable sources.
  6. Be tolerant. Everything that's good in nature comes from cooperation.
  Paraphrased a bit from Alberto Bertogli's Buena Onda License Agreement.
  For the full text, see: http://blitiri.com.ar/p/bola/


Changes
2009-10-04  Jonathan Yu <frequency@cpan.org>

	Grr, so many places with my email address :)

2009-10-04  Jonathan Yu <frequency@cpan.org>

	Fix email address

2009-10-03  Jonathan Yu <frequency@cpan.org>

	Change my mail address to jawnsy@cpan.org

2009-10-03  Jonathan Yu <frequency@cpan.org>

	Changes to MANIFEST.SKIP to accomodate the new file structure

2009-10-03  Jonathan Yu <frequency@cpan.org>

	Update 01build.t to test for files in the right place (check the build/ directory for compiled files)

2009-10-03  Jonathan Yu <frequency@cpan.org>

	Integrate version 1.00 of libjio

2009-08-19  Jonathan Yu <frequency@cpan.org>

	Added svn:ignore on intermediate build files

2009-08-19  Jonathan Yu <frequency@cpan.org>

	Small changes to My/Builder.pm subclass

2009-08-18  Jonathan Yu <frequency@cpan.org>

	Bump to version 1.002 and prepare for release

2009-08-18  Jonathan Yu <frequency@cpan.org>

	Call SUPER::ACTION_clean instead of ACTION_install when providing overrided clean action

2009-08-18  Jonathan Yu <frequency@cpan.org>

	- Remove modules available in core since Perl 5.6 from requires
	- Added ignore expressions for libjio's build-flags and libjio.so.0.90 (the symlink)
	- Override clean action to clean libjio files

2009-08-18  Jonathan Yu <frequency@cpan.org>

	- Call the README field 'LICENSE' rather than 'COPYRIGHT AND LICENCE' -- since we don't exactly have a copyright

2009-08-17  Jonathan Yu <frequency@cpan.org>

	Updated test sequence

2009-08-17  Jonathan Yu <frequency@cpan.org>

	Add a release date, prepare for release via PAUSE

2009-08-17  Jonathan Yu <frequency@cpan.org>

	Remove Perl::Critic from tests

2009-08-17  Jonathan Yu <frequency@cpan.org>

	Write a Changes file by hand

2009-08-17  Jonathan Yu <frequency@cpan.org>

	Note that Test::NoWarnings counts as 1 test

2009-08-17  Jonathan Yu <frequency@cpan.org>

	Changed to version 1.001 for upcoming release

2009-08-17  Jonathan Yu <frequency@cpan.org>

	Fix the nested SKIP block (prevents errors with plan/run mismatch)

2009-08-17  Jonathan Yu <frequency@cpan.org>

	- Remove the generated files (Makefile.PL, MANIFEST, META.yml)
	- Update MANIFEST.SKIP' paths -- Build dist was causing libjio's Makefile to be skipped, resulting in build problems

2009-08-17  Jonathan Yu <frequency@cpan.org>

	Changed README to UNIX format rather than DOS format

2009-08-17  Jonathan Yu <frequency@cpan.org>

	- Move the version test inside the SKIP block -- doesn't make sense to run it if it's not installed; this should fix the CPAN Testers errors (thanks hobbs@freenode #perl)
	- Switch version number to a simple string, per David Golden's Article -- http://www.dagolden.com/index.php/369/version-numbers-should-be-boring/
	- Bumped version to 1.0.5 (1.000005)
	- Remove dependency on version.pm

2009-08-17  Jonathan Yu <frequency@cpan.org>

	Updated MANIFEST with new libjio file names

2009-08-16  Jonathan Yu <frequency@cpan.org>

	[Integrating new upstream version] Updated to version 0.90 available online. TODO: integrate this in a script somehow

2009-08-16  Jonathan Yu <frequency@cpan.org>

	[Integrating new upstream version] libjio.skel.pc renamed to libjio.pc.in upstream

2009-08-16  Jonathan Yu <frequency@cpan.org>

	[Integrating new upstream version] Doxyfile.base renamed to Doxyfile.base.in upstream

2009-07-26  Jonathan Yu <frequency@cpan.org>

	- Fix core tests to correct regression in version 1.0.3
	- Bump version to 1.0.4

2009-07-25  Jonathan Yu <frequency@cpan.org>

	- Changed the method of probing @PATH for gmake, now we don't need to execute which
	- Bumped version number to 1.0.3

2009-07-25  Jonathan Yu <frequency@cpan.org>

	Added (awful) coverage information.

2009-07-25  Jonathan Yu <frequency@cpan.org>

	- Added some commands so that the void context paths are touched (increases coverage a bit)

2009-07-25  Jonathan Yu <frequency@cpan.org>

	- Various POD updates
	- Overhaul test scripts, remove author tests from Recommends (Alias' journal #5375)
	- Update LICENSE and README (no longer generate a readme, just use a simple one)

2009-07-01  Jonathan Yu <frequency@cpan.org>

	Revert the change to 02core.t which makes Env::Sanctify::Auto necessary, and remove taint checking for that test

2009-07-01  Jonathan Yu <frequency@cpan.org>

	- Bumped Alien::Libjio version to 1.0.2
	- Upgraded to new libjio 0.51 (released 2009-06-30)
	- There are no Ruby bindings, so remove that from the Build.PL prompt asking if they should be installed

2009-06-25  Jonathan Yu <frequency@cpan.org>

	Added feature to use gmake if available, and warn the user if on a BSD machine and it is unavailable

2009-06-25  Jonathan Yu <frequency@cpan.org>

	Added detection for BSD-ish OSes, so that gmake is preferred if available

2009-06-18  Jonathan Yu <frequency@cpan.org>

	Bumped to version 1.0.1

2009-06-14  Jonathan Yu <frequency@cpan.org>

	No longer enforce "one dot only" in filenames. This breaks compatibility with OpenVMS.

2009-06-14  Jonathan Yu <frequency@cpan.org>

	Updated perlcriticrc with new preferences, removed "no critic" stuff for those in Alien/Libjio.pm

2009-06-14  Jonathan Yu <frequency@cpan.org>

	- Added limitation warnings
	- Other minor documentation fixes

2009-06-13  Jonathan Yu <frequency@cpan.org>

	- A safer ENV{PATH}
	- Don't do ExtUtils::Liblist twice in t/02core.t

2009-06-12  Jonathan Yu <frequency@cpan.org>

	Scalar context doesn't do what it usually does; we should consider returning a string containing all flags (join'd)...

2009-06-12  Jonathan Yu <frequency@cpan.org>

	Fix skipped test count

2009-06-12  Jonathan Yu <frequency@cpan.org>

	Added an example, more tests, updated MANIFEST

2009-06-12  Jonathan Yu <frequency@cpan.org>

	Added libjio compiled files to MANIFEST.SKIP

2009-06-12  Jonathan Yu <frequency@cpan.org>

	Skip 3 tests inside SKIP block

2009-06-12  Jonathan Yu <frequency@cpan.org>

	Added t/00build.t - I'm still not sure what's causing the Kwalitee test failures though


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.2808_01
require 5.006;
    
    unless (eval "use Module::Build::Compat 0.02; 1" ) {
      print "This module requires Module::Build to install itself.\n";
      
      require ExtUtils::MakeMaker;
      my $yn = ExtUtils::MakeMaker::prompt
	('  Install Module::Build now from CPAN?', 'y');
      
      unless ($yn =~ /^y/i) {
	die " *** Cannot install without Module::Build.  Exiting ...\n";
      }
      
      require Cwd;
      require File::Spec;
      require CPAN;
      
      # Save this 'cause CPAN will chdir all over the place.
      my $cwd = Cwd::cwd();
      
      CPAN::Shell->install('Module::Build::Compat');
      CPAN::Shell->expand("Module", "Module::Build::Compat")->uptodate
	or die "Couldn't install Module::Build, giving up.\n";
      
      chdir $cwd or die "Cannot chdir() back to $cwd: $!";
    }
    eval "use Module::Build::Compat 0.02; 1" or die $@;
    use lib 'inc';
    Module::Build::Compat->run_build_pl(args => \@ARGV);
    require My::Builder;
    Module::Build::Compat->write_makefile(build_class => 'My::Builder');


inc/My/Builder.pm
# My::Builder
#  A local Module::Build subclass for installing libjio
#
# $Id: Builder.pm 8670 2009-08-20 02:38:50Z FREQUENCY@cpan.org $

package My::Builder;

use strict;
use warnings;

use base 'Module::Build';

use Cwd ();
use Carp ();

my $ORIG_DIR = Cwd::cwd();

# These are utility commands for getting into and out of our build directory
sub _chdir_or_die {
  use File::Spec ();
  my $dir = File::Spec->catfile(@_);
  chdir $dir or Carp::croak("Failed to chdir to $dir: $!");
}
sub _chdir_back {
  chdir $ORIG_DIR or Carp::croak("Failed to chdir to $ORIG_DIR: $!");
}

sub ACTION_code {
  my ($self) = @_;

  my $rc = $self->SUPER::ACTION_code;
  if ($self->notes('build_libjio')) {
    # Get into our build directory; either libjio (all) or libjio/libjio
    # (bindings only)
    if ($self->notes('extra')) {
      _chdir_or_die('libjio');
    }
    else {
      _chdir_or_die('libjio', 'libjio');
    }

    # Run the make system to do the rest, but save the return code
    system($self->notes('make'));
    $rc = $? >> 8;

    # Make sure we change the directory back before adding notes, or they
    # won't persist (in _build state)
    _chdir_back();
    $self->notes(build_result => $rc);
  }

  return $rc;
}

sub ACTION_install {
  my ($self) = @_;

  my $rc = $self->SUPER::ACTION_install;
  if ($self->notes('build_libjio')) {
    # Get into our build directory
    if ($self->notes('extra')) {
      _chdir_or_die('libjio');
    }
    else {
      _chdir_or_die('libjio', 'libjio');
    }

    # Run the make system to do the rest
    $rc = (system($self->notes('make'), 'install') == 0) ? 1 : 0;
    _chdir_back();
  }

  return $rc;
}

sub ACTION_clean {
  my ($self) = @_;

  my $rc = $self->SUPER::ACTION_clean;
  _chdir_or_die('libjio');
  $rc = (system($self->notes('make'), 'clean') == 0) ? 1 : 0;
  _chdir_back();

  return $rc;
}

1;


lib/Alien/Libjio.pm
# Alien::Libjio
#  A Perl package to install libjio, a library for Journalled I/O.
#
# $Id: Libjio.pm 9520 2009-10-04 03:58:25Z FREQUENCY@cpan.org $

package Alien::Libjio;

use strict;
use warnings;
use Carp ();

=head1 NAME

Alien::Libjio - Utility package to install and locate libjio

=head1 VERSION

Version 1.003 ($Id: Libjio.pm 9520 2009-10-04 03:58:25Z FREQUENCY@cpan.org $)

=cut

our $VERSION = '1.003';
$VERSION = eval $VERSION;

=head1 DESCRIPTION

To ensure reliability, some file systems and databases provide support for
something known as journalling. The idea is to ensure data consistency by
creating a log of actions to be taken (called a Write Ahead Log) before
committing them to disk. That way, if a transaction were to fail due to a
system crash or other unexpected event, the write ahead log could be used to
finish writing the data.

While this functionality is often available with networked databases, it can
be a rather memory- and processor-intensive solution, even where reliable
writes are important. In other cases, the filesystem does not provide native
journalling support, so other tricks may be used to ensure data integrity,
such as writing to a separate temporary file and then overwriting the file
instead of modifying it in-place. Unfortunately, this method cannot handle
threaded operations appropriately.

Thankfully, Alberto Bertogli published a userspace C library called libjio
that can provide these features in a small (less than 1500 lines of code)
library with no external dependencies.

This package is designed to install it, and provide a way to get the flags
necessary to compile programs using it. It is particularly useful for Perl XS
programs that use it, such as B<IO::Journal>.

=head1 SYNOPSIS

  use Alien::Libjio;

  my $jio = Alien::Libjio->new;
  my $ldflags = $jio->ldflags;
  my $cflags = $jio->cflags;

=head1 COMPATIBILITY

This module was tested under Perl 5.10.0, using Debian Linux. However, because
it's Pure Perl and doesn't do anything too obscure, it should be compatible
with any version of Perl that supports its prerequisite modules.

By default, this library is installed wherever the main system libraries are
usually installed. As a result, C<Alien::Libjio> will only work if installed
with root permissions.

If you encounter any problems on a different version or architecture, please
contact the maintainer.

=head1 METHODS

=head2 Alien::Libjio->new

Creates a new C<Alien::Libjio> object, which essentially just has a few
convenience methods providing useful information like compiler and linker
flags.

Example code:

  my $jio = Alien::Libjio->new();

This method will return an appropriate B<Alien::Libjio> object or throw an
exception on error.

=cut

sub new {
  my ($class, @seed) = @_;

  Carp::croak('You must call this as a class method') if ref($class);

  my $self = {
    installed => 0,
  };

  bless($self, $class);

  $self->_try_pkg_config()
    or $self->_try_liblist()
    or delete($self->{method});

  return $self;
}

=head2 $jio->installed

Determine if a valid installation of libjio has been detected in the system.
This method will return a true value if it is, or undef otherwise.

Example code:

  print "okay\n" if $jio->installed;

=cut

sub installed {
  my ($self) = @_;

  Carp::croak('You must call this method as an object') unless ref($self);

  return $self->{installed};
}

=head2 $jio->version

Determine the installed version of libjio, as a string.

Currently versions are simply floating-point numbers, so you can treat the
version number as such, but this behaviour is subject to change.

Example code:

  my $version = $jio->version;

=cut

sub version {
  my ($self) = @_;

  Carp::croak('You must call this method as an object') unless ref($self);

  return $self->{version};
}

=head2 $jio->ldflags

=head2 $jio->linker_flags

This returns the flags required to link C code with the local installation of
libjio (typically in the LDFLAGS variable). It is particularly useful for
building and installing Perl XS modules such as L<IO::Journal>.

In scalar context, it returns an array reference suitable for passing to
other build systems, particularly L<Module::Build>. In list context, it gives
a normal array so that C<join> and friends will work as expected.

Example code:

  my $ldflags = $jio->ldflags;
  my @ldflags = @{ $jio->ldflags };
  my $ldstring = join(' ', $jio->ldflags);
  # or:
  # my $ldflags = $jio->linker_flags;

=cut

sub ldflags {
  my ($self) = @_;

  Carp::croak('You must call this method as an object') unless ref($self);

  # Return early if called in void context
  return unless defined wantarray;

  # If calling in array context, dereference and return
  return @{ $self->{ldflags} } if wantarray;

  return $self->{ldflags};
}

# Glob to create an alias to ldflags
*linker_flags = *ldflags;

=head2 $jio->cflags

=head2 $jio->compiler_flags

This method returns the compiler option flags to compile C code which uses
the libjio library (typically in the CFLAGS variable). It is particularly
useful for building and installing Perl XS modules such as L<IO::Journal>.

Example code:

  my $cflags = $jio->cflags;
  my @cflags = @{ $jio->cflags };
  my $ccstring = join(' ', $jio->cflags);
  # or:
  # my $cflags = $jio->compiler_flags;

=cut

sub cflags {
  my ($self) = @_;

  Carp::croak('You must call this method as an object') unless ref($self);

  # Return early if called in void context
  return unless defined wantarray;

  # If calling in array context, dereference and return
  return @{ $self->{cflags} } if wantarray;

  return $self->{cflags};
}
*compiler_flags = *cflags;

=head2 $jio->method

=head2 $jio->how

This method returns the method the module used to find information about
libjio. The following methods are currently used (in priority order):

=over

=item *

pkg-config: the de-facto package information tool

=item *

ExtUtils::Liblist: a utility module used by ExtUtils::MakeMaker

=back

Example code:

  if ($jio->installed) {
    print 'I found this information using: ', $jio->how, "\n";
  }

=cut

sub method {
  my ($self) = @_;

  Carp::croak('You must call this method as an object') unless ref($self);

  return $self->{method};
}
*how = *method;

# Private methods to find & fill out information

use IPC::Open3 ('open3');

sub _get_pc {
  my ($key) = @_;

  my $read;
  my $pid = open3(undef, $read, undef, 'pkg-config', 'libjio', '--' . $key);
  # We're using blocking wait, so the return value doesn't matter
  ## no critic(RequireCheckedSyscalls)
  waitpid($pid, 0);

  # Check the exit status; 0 = success - nonzero = failure
  if (($? >> 8) == 0) {
    # The value we got back
    return <$read>;
  }
  return (undef, <$read>) if wantarray;
  return;
}

sub _try_pkg_config {
  my ($self) = @_;

  my ($value, $err) = _get_pc('cflags');
  return unless (defined $value && length $value);
  #if (defined $err && length $err) {
  #  #warn "Problem with pkg-config; using ExtUtils::Liblist instead\n";
  #  return;
  #}

  $self->{installed} = 1;
  $self->{method} = 'pkg-config';

  # pkg-config returns things with a newline, so remember to remove it
  $self->{cflags} = [ split(' ', $value) ];
  $self->{ldflags} = [ split(' ', _get_pc('libs')) ];
  $self->{version} = _get_pc('modversion');

  return 1;
}

sub _try_liblist {
  my ($self) = @_;

  use ExtUtils::Liblist ();
  local $SIG{__WARN__} = sub { }; # mask warnings

  my (undef, undef, $ldflags, $ldpath) = ExtUtils::Liblist->ext('-ljio');
  return unless (defined($ldflags) && length($ldflags));

  $self->{installed} = 1;
  $self->{method} = 'ExtUtils::Liblist';

  # Empty out cflags; initialize it
  $self->{cflags} = [];

  my $read;
  my $pid = open3(undef, $read, undef, 'getconf', 'LFS_CFLAGS');

  # We're using blocking wait, so the return value doesn't matter
  ## no critic(RequireCheckedSyscalls)
  waitpid($pid, 0);

  # Check the status code
  if (($? >> 8) == 0) {
    # This only takes the first line
    push(@{ $self->{cflags} }, split(' ', <$read>));
  }
  else {
    warn 'Problem using getconf: ', <$read>, "\n";
    push(@{ $self->{cflags} },
      '-D_LARGEFILE_SOURCE',
      '-D_FILE_OFFSET_BITS=64',
    );
  }

  # Used for resolving the include path, relative to lib
  use Cwd ();
  use File::Spec ();
  push(@{ $self->{cflags} },
    # The include path is taken as: $libpath/../include
    '-I' . Cwd::realpath(File::Spec->catfile(
      $ldpath,
      File::Spec->updir(),
      'include'
    ))
  );

  push(@{ $self->{ldflags} },
    '-L' . $ldpath,
    $ldflags,
  );

  return 1;
}

=head1 AUTHOR

Jonathan Yu E<lt>jawnsy@cpan.orgE<gt>

=head2 CONTRIBUTORS

Your name here ;-)

=head1 ACKNOWLEDGEMENTS

=over

=item *

Special thanks to Alberto Bertogli E<lt>albertito@blitiri.com.arE<gt> for
developing this useful library and for releasing it into the public domain.

=back

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Alien::Libjio

You can also look for information at:

=over

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Alien-Libjio>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Alien-Libjio>

=item * Search CPAN

L<http://search.cpan.org/dist/Alien-Libjio>

=item * CPAN Request Tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-Libjio>

=item * CPAN Testing Service (Kwalitee Tests)

L<http://cpants.perl.org/dist/overview/Alien-Libjio>

=item * CPAN Testers Platform Compatibility Matrix

L<http://cpantesters.org/show/Alien-Libjio.html>

=back

=head1 REPOSITORY

You can access the most recent development version of this module at:

L<http://svn.ali.as/cpan/trunk/Alien-Libjio>

If you are a CPAN developer and would like to make modifications to the code
base, please contact Adam Kennedy E<lt>adamk@cpan.orgE<gt>, the repository
administrator. I only ask that you contact me first to discuss the changes you
wish to make to the distribution.

=head1 FEEDBACK

Please send relevant comments, rotten tomatoes and suggestions directly to the
maintainer noted above.

If you have a bug report or feature request, please file them on the CPAN
Request Tracker at L<http://rt.cpan.org>. If you are able to submit your bug
report in the form of failing unit tests, you are B<strongly> encouraged to do
so.

=head1 SEE ALSO

L<IO::Journal>, a Perl module that provides an interface to libjio.

L<http://blitiri.com.ar/p/libjio/>, Alberto Bertogli's page about libjio,
which explains the purpose and features of libjio.

=head1 CAVEATS

=head2 KNOWN BUGS

There are no known bugs as of this release.

=head2 LIMITATIONS

=over

=item *

This module can only search known/common paths for libjio installations.
It does try to use B<pkg-config> and B<ExtUtils::Liblist> to find a libjio
installation on your system, but it cannot predict where files might have
been installed. As a result, this package might install a duplicate copy
of libjio.

=item  *

There is currently no way to save a custom library installation path for
libjio. This is likely to change in the future.

=item *

B<pkg-config> may fail if you have an insecure $ENV{PATH} variable. Due to
the way IPC::Open3 works, taintedness exceptions are suppressed and pkg-config
seems to fail for no reason. The recommended fix for this is to use a module
like L<Env::Sanctify::Auto> or to otherwise clean up the calling environment.
Another workaround is to disable taint checking, but that's not recommended.
(See: L<http://rt.perl.org/rt3/Ticket/Display.html?id=66572>)

=back

=head1 LICENSE

In a perfect world, I could just say that this package and all of the code
it contains is Public Domain. It's a bit more complicated than that; you'll
have to read the included F<LICENSE> file to get the full details.

=head1 QUALITY ASSURANCE METRICS

=head2 TEST COVERAGE

  ----------------------- ------ ------ ------ ------ ------ ------
  File                     stmt   bran   cond   sub    pod   total
  ----------------------- ------ ------ ------ ------ ------ ------
  Alien/Libjio.pm         97.7   84.4   66.7   100.0  100.0  93.4

Okay, granted, the coverage sucks -- generally, I aim for 100% in all of
the categories, using fault injection to test all the code paths. Due to
the nature of the module it's somewhat difficult to do.

=head1 DISCLAIMER OF WARRANTY

The software is provided "AS IS", without warranty of any kind, express or
implied, including but not limited to the warranties of merchantability,
fitness for a particular purpose and noninfringement. In no event shall the
authors or copyright holders be liable for any claim, damages or other
liability, whether in an action of contract, tort or otherwise, arising from,
out of or in connection with the software or the use or other dealings in
the software.

=cut

1;


Build.PL
#!/usr/bin/perl

# Build.PL
#  Script to build and install this distribution
#
# $Id: Build.PL 9520 2009-10-04 03:58:25Z FREQUENCY@cpan.org $

use strict;
use warnings;

use lib 'inc';
use My::Builder;

use Env '@PATH';

my $builder = My::Builder->new(
  module_name           => 'Alien::Libjio',
  license               => 'unrestricted',
  dist_author           => 'Jonathan Yu <jawnsy@cpan.org>',
  dist_version_from     => 'lib/Alien/Libjio.pm',
  dynamic_config        => 1,
  create_readme         => 0,
  recursive_test_files  => 1,
  sign                  => 1,
  create_packlist       => 1,

  # Maintain compatibility with ExtUtils::MakeMaker installations
  create_makefile_pl    => 'passthrough',

  requires => {
    'perl'              => 5.006,
  },
  build_requires => {
    # User tests for good functionality
    'Test::More'              => 0.62,
    'Test::NoWarnings'        => 0.084,

    # For the C compiling process
    'ExtUtils::CBuilder'      => 0,
  },
  recommends => {
  },
  conflicts => {
  },

  add_to_cleanup => [ 'Alien-Libjio-*' ],
  script_files => [],

  meta_merge => {
    resources => {
      # Custom resources (must begin with an uppercase letter)
      Ratings      => 'http://cpanratings.perl.org/d/Alien-Libjio',

      # Official keys (homepage, license, bugtracker)
      repository   => 'http://svn.ali.as/cpan/trunk/Alien-Libjio',
      bugtracker   => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-Libjio',
      license      => 'http://edwardsamuels.com/copyright/beyond/articles/public.html',
    },
  },
);

# Use Alien::Libjio to see if it's already installed
use lib 'lib';
use Alien::Libjio;

my $jio = Alien::Libjio->new();

unless ($jio->installed) {
  # Ask the user if they'd like to install this; if not, then exit
  $builder->y_n('libjio was not found on your system. Install it now?', 'y')
    or exit;

  $builder->notes(build_libjio => 1);

  # Ask the user what 'make' program to invoke
  my $make;
  if (exists($ENV{MAKE}) && length($ENV{MAKE})) {
    $make = $ENV{MAKE};
  }
  else {
    use Config '%Config';
    $make = $Config{make};

    # Probe for GNU Make (useful on BSD/Unix variants)
    if ($make eq 'make' && grep { -x $_ . '/gmake' } @PATH) {
      $make = 'gmake';
      if ($^O =~ /bsd$/) {
        print {*STDERR} "warning: your system is a BSD variant but " .
          "gmake wasn't found.\n";
      }
    }
  }
  $make = $builder->prompt('What is your system "make" command?', $make);

  $builder->notes(make => $make);

  # Figure out if we should do a full install
  my $extra = $builder->y_n('libjio includes other files including Python ' .
    'bindings. Install them too?', 'n');
  $builder->notes(extra => $extra);
}

$builder->create_build_script();
